Make ciphersuite function implementation explicit (including input validation) (#360)
cfrg/draft-irtf-cfrg-voprf: draft-irtf-cfrg-voprf.md

Applications MUST check that input Element values received over the wire
are not the group identity element. This check is handled after deserializing

> Element values; see ciphersuite}} for more information and requirements

Reviewer: are these some extra `}}` or am I missing something?

    DST = "HashToScalar-" || contextString, and output length 64, interpret
    `uniform_bytes` as a 512-bit integer in little-endian order, and reduce the
    integer modulo `Group.Order()`.
  - ScalarInverse(s): Returns the multipicative inverse of input Scalar `s` mod `Group.Order()`.
  - RandomScalar(): Implemented by returning a uniformly random Scalar in the range

>     \[0, `G.Order()` - 1\]. Refer to {{random-scalar}} for implementation guidance.

Reviewer 2: I consider these two functions (ScalarInverse and RandomScalar) are generic over each group instantiation. So we might remove them from each group instance.

    DST = "HashToScalar-" || contextString, and output length 64, interpret
    `uniform_bytes` as a 512-bit integer in little-endian order, and reduce the
    integer modulo `Group.Order()`.

>   - ScalarInverse(s): Returns the multipicative inverse of input Scalar `s` mod `Group.Order()`.

Reviewer 2: ```suggestion
  - ScalarInverse(s): Returns the multiplicative inverse of input Scalar `s` mod `Group.Order()`.
```

  - SerializeElement(A): Implemented using the 'Encode' function from Section 5.3.2 of {{!RISTRETTO}}; Ne = 56.
  - DeserializeElement(buf): Implemented using the 'Decode' function from Section 5.3.1 of {{!RISTRETTO}}.
    Additionally, this function validates that the resulting element is not the group
    identity element. If these checks fail, deserialization returns an InputValidationError error.
  - SerializeScalar(s): Implemented by outputting the little-endian 56-byte encoding of
    the Scalar value; Ns = 56.
  - DeserializeScalar(buf): Implemented by attempting to deserialize a Scalar from a
    little-endian 56-byte string. This function can fail if the input does not
    represent a Scalar in the range \[0, `G.Order()` - 1\].
- Hash: SHAKE-256; Nh = 64.
- ID: 0x0002

### OPRF(P-256, SHA-256)

- Group: P-256 (secp256r1) {{NISTCurves}}

>   - Order(): Return 0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551.

Reviewer 2: For group order, be consistent about using hex or decimal numbers.

Author: Let's stick with this as-is for now. We can update later if needed.

  - Identity(): As defined in {{NISTCurves}}.
  - Generator(): As defined in {{NISTCurves}}.
  - RandomScalar(): Implemented by returning a uniformly random Scalar in the range
    \[0, `G.Order()` - 1\]. Refer to {{random-scalar}} for implementation guidance.
  - HashToGroup(): Use hash_to_curve with suite P256_XMD:SHA-256_SSWU_RO\_
    {{!I-D.irtf-cfrg-hash-to-curve}} and DST =
    "HashToGroup-" || contextString.
  - HashToScalar(): Use hash_to_field from {{!I-D.irtf-cfrg-hash-to-curve}}
    using L = 48, `expand_message_xmd` with SHA-256,
    DST = "HashToScalar-" || contextString, and
    prime modulus equal to `Group.Order()`.
  - ScalarInverse(s): Returns the multipicative inverse of input Scalar `s` mod `Group.Order()`.
  - SerializeElement(A): Implemented using the compressed Elliptic-Curve-Point-to-Octet-String
    method according to {{SEC1}};  Ne = 33.
  - DeserializeElement(buf): Implemented by attempting to deserialize a 33 byte input string to
    a public key using the compressed Octet-String-to-Elliptic-Curve-Point method according to {{SEC1}},
    and then performs partial public-key validation as defined in section 5.6.2.3.4 of

>     {{!KEYAGREEMENT=DOI.10.6028/NIST.SP.800-56Ar3}}. This includes checking that the
>     coordinates of the resulting point are in the correct range, that the point is on
>     the curve, and that the point is not the point at infinity. Additionally, this function
>     validates that the resulting element is not the group identity element.

Reviewer 2: ```suggestion
    {{!KEYAGREEMENT=DOI.10.6028/NIST.SP.800-56Ar3}}. This includes checking that the
    coordinates of the resulting point are in the correct range, that the point is on
    the curve, and that the point is not the group identity element.
```

These sentences are redundant.

    input does not represent a Scalar in the range \[0, `G.Order()` - 1\].
- Hash: SHA-256; Nh = 32.
- ID: 0x0003

### OPRF(P-384, SHA-384)

- Group: P-384 (secp384r1) {{NISTCurves}}
  - Order(): Return 0xffffffffffffffffffffffffffffffffffffffffffffffffc7634d81f4372ddf581a0db248b0a77aecec196accc52973.
  - Identity(): As defined in {{NISTCurves}}.
  - Generator(): As defined in {{NISTCurves}}.
  - RandomScalar(): Implemented by returning a uniformly random Scalar in the range
    \[0, `G.Order()` - 1\]. Refer to {{random-scalar}} for implementation guidance.
  - HashToGroup(): Use hash_to_curve with suite P384_XMD:SHA-384_SSWU_RO\_
    {{!I-D.irtf-cfrg-hash-to-curve}} and DST =
    "HashToGroup-" || contextString.
  - HashToScalar(): Use hash_to_field from {{!I-D.irtf-cfrg-hash-to-curve}}
    using L = 72, `expand_message_xmd` with SHA-384,
    DST = "HashToScalar-" || contextString, and
    prime modulus equal to `Group.Order()`.
  - ScalarInverse(s): Returns the multipicative inverse of input Scalar `s` mod `Group.Order()`.
  - SerializeElement(A): Implemented using the compressed Elliptic-Curve-Point-to-Octet-String
    method according to {{SEC1}}; Ne = 49.

>   - DeserializeElement(buf): Implemented by attempting to deserialize a 49 byte input string to

Reviewer 2: We had normalized the document to use `byte array` or `array of bytes`, instead of `byte string`.

```suggestion
  - DeserializeElement(buf): Implemented by attempting to deserialize a 49-byte array  to
```

validation is done for DeserializeElement and DeserializeScalar.

Additionally, future ciphersuites must take care when choosing the
security level of the group. See {{limits}} for additional details.

## Random Scalar Generation {#random-scalar}

Two popular algorithms for generating a random integer uniformly distributed in
the range \[0, G.Order() -1\] are as follows:

### Rejection Sampling

Generate a random byte array with `Ns` bytes, and attempt to map to a Scalar
by calling `DeserializeScalar` in constant time. If it succeeds, return the
result. If it fails, try again with another random byte array, until the
procedure succeeds. Failure to implement `DeserializeScalar` in constant time
can leak information about the underlying corresponding Scalar.

As an optimization, if the group order is very close to a power of
2, it is acceptable to omit the rejection test completely.  In
particular, if the group order is p, and there is an integer b
such that `p - 2<sup>b</sup>| < 2<sup>(b/2)</sup>`, then
`RandomScalar` can simply return a uniformly random integer of at
most b bits.

### Random Number Generation Using Extra Random Bits

> Generate a random byte array with `l = ceil(((3 * ceil(log2(G.Order()))) / 2) / 8)`

Author: ```suggestion
Generate a random byte array with `ceil(((3 * ceil(log2(G.Order()))) / 2) / 8)`
```

> Generate a random byte array with `l = ceil(((3 * ceil(log2(G.Order()))) / 2) / 8)`
> bytes, and interpret it as an integer; reduce the integer modulo `G.Order()` and return the
> result. See {{I-D.irtf-cfrg-hash-to-curve, Section 5}} for the underlying derivation of `l`.

Author: ```suggestion
Generate a random byte array with `L = ceil(((3 * ceil(log2(G.Order()))) / 2) / 8)`
bytes, and interpret it as an integer; reduce the integer modulo `G.Order()` and return the
result. See {{I-D.irtf-cfrg-hash-to-curve, Section 5}} for the underlying derivation of `L`.
```

must adhere to the implementation and security considerations discussed
in {{!I-D.irtf-cfrg-hash-to-curve}} when instantiating the function.

The DeserializeElement and DeserializeScalar functions instantiated for a
particular prime-order group corresponding to a ciphersuite MUST adhere to
the description in {{pog}}. Future ciphersuites MUST describe how input
validation is done for DeserializeElement and DeserializeScalar.

Additionally, future ciphersuites must take care when choosing the
security level of the group. See {{limits}} for additional details.

## Random Scalar Generation {#random-scalar}

Two popular algorithms for generating a random integer uniformly distributed in
the range \[0, G.Order() -1\] are as follows:

> ### Rejection Sampling

Reviewer 2: Section 5 of Hash to curve draft states:

```
   Implementors MUST NOT use rejection sampling to generate a uniformly
   random element of F, to ensure that the hash_to_field function is
   amenable to constant-time implementation.  The reason is that
   rejection sampling procedures are difficult to implement in constant
   time, and later well-meaning "optimizations" may silently render an
   implementation non-constant-time.
```

how this section in the voprf draft is in accordance to that recommendation?

Generate a random byte array with `Ns` bytes, and attempt to map to a Scalar
by calling `DeserializeScalar` in constant time. If it succeeds, return the
result. If it fails, try again with another random byte array, until the
procedure succeeds. Failure to implement `DeserializeScalar` in constant time
can leak information about the underlying corresponding Scalar.

As an optimization, if the group order is very close to a power of
2, it is acceptable to omit the rejection test completely.  In
particular, if the group order is p, and there is an integer b
such that `p - 2<sup>b</sup>| < 2<sup>(b/2)</sup>`, then
`RandomScalar` can simply return a uniformly random integer of at

> most b bits.

Reviewer 2: and reduce `mod p`?
in other words, if after sampling one gets a number above `Order() <= s < 2^b`, is this a valid scalar?

Author: Reduction is unnecessary here since it'll be less than p (by definition).

> ### Wide Reduction

Reviewer 2: ```suggestion
### Random Number Generation Using Extra Random Bits
```

* Similar to Appendix B.4.1 of FIPS 186-4.

> Generate a random byte array with `l = ceil(((3 * ceil(log2(G.Order()))) / 2) / 8)`

Reviewer 2: double check, does this `l` refers to `L` in has to curve draft? 
If so, how the security parameter is involved in this instantiation?

```suggestion
Generate a random byte array with `l = ceil(((3 * ceil(log2(G.Order()))) / 2) / 8)`
```
